File: cpp/2sum.cpp
#include <iostream>
#include <vector>

/*
2 SUM
2 NUMBERS THAT ADD TO THE TARGET

clang++ 2sum.cpp -o linout/2sum
linout/2sum

*/

class Solution {
public:
    static std::vector<int> twoSum(const std::vector<int>& nums, const int& target) {
        for(int i=0; i<nums.size(); ++i){
            int diff = target-nums[i];
            for(int z=i+1; z<nums.size(); ++z){
                if(nums[z]==diff) return {i, z};
            }
            
        }
        //default
        return {0,0};
        
    }
};

int main(){
    std::cout << '{';
    for(int i: Solution::twoSum({2, 7}, 9)){
        std::cout << i << ',';
    }
    std::cout<<'}';
}

File: python/2sum.py

import pandas as pd
from pandas import DataFrame as df

'''

MOST POPULAR LEETCODE PROBLEM
2SUM

given an array of ints, return indices of 2 numbers that add to a number

theres always 1 solution
check every combination of 2 values and see if they sum to the target

brute force algorithm
time complexity O(n^2)
n*n

implement a hashmap or a dictionary
assigning the difference of target - index to the index


'''

class Solution:
    def two_sum(self, input:list[int], target:int) -> list[int]:
        prevMap = df({})
        
        for i, n in enumerate(input):
            #NOTE: enumerate returns the val and its index
            # i is index, n is the number in that index
            # if only 1 arg itll return a list of both
            diff = target- n
            if diff in prevMap.columns: #if the difference exists in columns
                return [prevMap.columns.get_loc(diff), i]
            prevMap[n] = i

            

print(Solution.two_sum(Solution, [1,4,7,2], 6))




